### PR TITLE
e2e: Migrate dynamic fees test from kurtosis

### PR DESCRIPTION
**Pull request by marun**
## Why this should be merged

Fulfills one of the requirements for #1547 (migration of Kurtosis tests)

## How this works

- reimplements avalanche testing's [dynamic fees test](https://github.com/ava-labs/avalanche-testing/blob/master/tests/dynfees/dynfees.go)
- enables use of a test-private network (possible now that the shared network is only 2 nodes) to avoid having other tests interfere with gas pricing

## How this was tested

Ran ./scripts/test.e2e.sh

## TODO

- [x] Merge the parent PR (#1941 )

### CODE REVIEW

File: tests/e2e/c/hashing_contract.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// AUTOMATICALLY GENERATED. DO NOT EDIT!
+// Generated from hashing.sol by compile-contract.sh
```
Comment (danlaine) [inline tests/e2e/c/hashing_contract.go:5]: Can we add `compile-contract.sh` to this directory?

**Review by danlaine — COMMENTED** [review]
1 nit otherwise looks OK

File: tests/e2e/c/hashing_contract.go (hunk 1 above)
Comment (aaronbuchwald) [reply tests/e2e/c/hashing_contract.go:5]: We supported this until go-ethereum removed their Solidity compiler package. I'm not opposed to installing solc and using that where necessary.

I don't have a strong opinion on whether it's better to install solc and compile on-demand or hardcode and am open to either.

**Review by danlaine — APPROVED** [review]

File: tests/e2e/c/dynamic_fees.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/ava-labs/avalanchego/tests"
+	"github.com/ava-labs/avalanchego/tests/e2e"
+	"github.com/ava-labs/avalanchego/tests/fixture/testnet"
+	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
+)
+
+// This test uses the compiled bin for `hashing.sol` as
+// well as its ABI contained in `hashing_contract.go`.
+
+// This test needs to be run serially to minimize the potential for
```
[the remark this answers is not in the document]
Comment (danlaine) [reply tests/e2e/c/dynamic_fees.go:30]: Down to remove this :+1:

File: tests/e2e/e2e.go [hunk 3]
```diff
@@ -54,6 +56,16 @@ const (
 	// current time for validator addition to succeed, and adding 10
 	// seconds provides a buffer in case of any delay in processing.
 	DefaultValidatorStartTimeDiff = executor.SyncBound + 10*time.Second
+
+	DefaultGasLimit = uint64(21000) // Standard gas limit
```
Comment (hexfusion) [inline tests/e2e/e2e.go:60]: :+1:

File: tests/e2e/e2e.go [hunk 4]
```diff
@@ -144,6 +158,20 @@ func (te *TestEnvironment) NewEthClient(nodeURI testnet.NodeURI) ethclient.Clien
...
+// Create a new private network that is not shared with other tests.
+func (te *TestEnvironment) NewPrivateNetwork() testnet.Network {
+	// Load the shared network to retrieve its path and exec path
+	sharedNetwork, err := local.ReadNetwork(te.NetworkDir)
+	te.require.NoError(err)
+
+	// The private networks dir is under the shared network dir to ensure it
+	// will be included in the artifact uploaded in CI.
+	privateNetworksDir := filepath.Join(sharedNetwork.Dir, PrivateNetworksDirName)
+	te.require.NoError(os.MkdirAll(privateNetworksDir, perms.ReadWriteExecute))
```
Comment (hexfusion) [inline tests/e2e/e2e.go:170]: optional: I prefer to be explicit with error handling 
```suggestion
        err = os.MkdirAll(privateNetworksDir, perms.ReadWriteExecute)
	te.require.NoError(err)
```

File: tests/e2e/e2e.go [hunk 5]
```diff
@@ -259,3 +287,32 @@ func CheckBootstrapIsPossible(network testnet.Network) {
...
+			},
+		},
+		testnet.DefaultNodeCount,
+		testnet.DefaultFundedKeyCount,
+	)
+	require.NoError(err)
+	ginkgo.DeferCleanup(func() {
+		tests.Outf("Shutting down network\n")
+		require.NoError(network.Stop())
+	})
```
Comment (hexfusion) [inline tests/e2e/e2e.go:313]: feature request: disable this step with flag or ENV

Comment (hexfusion) [reply tests/e2e/e2e.go:313]: having it optional by ENV seems non intrusive and useful for debug, just for future considerations :)

**Review by hexfusion — APPROVED** [review]

**Pull request merged**